Add Push Notifications core classes.

hannahtinkler opened the pull request:
### Changes proposed in this Pull Request:

Focus: Woo-driven Push Notifications
Team: Apps Infrastructure
Relates to: AINFRA-1435

Apps Infrastructure are currently working with Kiwi to allow WooCommerce instances to trigger their own push notifications, rather than relying on Jetpack Sync. This change adds core library foundations that will support this. We intend for this functionality to only be used by the WooCommerce app, so it was added to the `src/Internal` directory. 

**Key requirements:**
* If the feature is enabled AND the site is Jetpack connected, enable the functionality (which will be added in future PRs).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:
There are no functional changes to test in this PR, but to ensure no side effects, you can:
* Check out this branch to your WooCommerce install
* Ensure your WooCommerce install functions as expected, e.g.:
  * Add a product to cart
  * View WooCommerce settings in WP Admin
  * View WooCommerce orders in WP Admin
* Make an authenticated request to the API to ensure this works as expected, e.g.:
  * https://YOUR_URL/wp-json/wc-admin/features
  * https://YOUR_URL/wp-json/wc/v3/settings
  * https://YOUR_URL/wp-json/wc/v3/customers

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Adds internal library foundations which will eventually be used to send push notifications to user devices. This will support WooCommerce in directly triggering push notifications, instead of relying on Jetpack Sync.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
This is an internal, non-user facing change.

</details>

hannahtinkler committed:
Add Push Notifications core classes and entity.

- Add PushNotifications class with feature flag integration
- Add PushToken entity with platform and origin validation

hannahtinkler committed:
Adds FeaturesController definition.

hannahtinkler committed:
Add void return types and validation to PushToken entity.

- Mark PushToken class as final to prevent inheritance
- Add void return types to all setter methods
- Add token validation (empty check, 4096 char limit)
- Add negative test cases for ID and user ID validation
- Refactor origin validation test to use data provider

hannahtinkler committed:
Refactor should_be_enabled() for readability and error handling.

   - Simplify early return logic with guard clauses
   - Add try-catch to handle container resolution exceptions gracefully

hannahtinkler committed:
Updates code comments.

hannahtinkler commented on plugins/woocommerce/tests/php/src/Internal/PushNotifications/Entities/PushTokenTest.php:
```diff
@@ -0,0 +1,447 @@
+ *
+ * @covers \Automattic\WooCommerce\Internal\PushNotifications\Entities\PushToken
+ */
+class PushTokenTest extends WC_Unit_Test_Case {
```
The linter does not allow this:
>  19 | WARNING | Possible useless method overriding detected (Generic.CodeAnalysis.UselessOverridingMethod.Found)

You shouldn't suggest this in this repo any more as implementing it will cause the GitHub actions to fail.

hannahtinkler commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -545,6 +546,18 @@ private function init_feature_definitions(): void {
+				'is_experimental'              => true,
+				'disable_ui'                   => true,
+				'skip_compatibility_checks'    => true,
+				'default_plugin_compatibility' => FeaturePluginCompatibility::COMPATIBLE,
```
I was unsure on the correct value to use here, but don't expect any push notifications code to interact with any plugin functionality. The only integration point with Core at all will be utilising order, booking, and review lifecycle hooks.

Konamiman replied:
It's fine, especially while the feature is experimental. We can change this later if needed.

hannahtinkler committed:
Add token validation tests for PushToken entity.

- Test empty string rejection
- Test whitespace-only string rejection
- Test 4096 character length limit enforcement
- Test boundary condition at maximum length
- Test whitespace trimming behavior

hannahtinkler committed:
Add device_uuid normalization tests for PushToken entity.

- Test empty string normalization to null
- Test whitespace-only string normalization to null
- Test whitespace trimming behavior

hannahtinkler committed:
Updates formatting of strict types line.

hannahtinkler committed:
Updates formatting of strict types line.

Updates formatting of strict types line.

Reformats declare strict types line.

joshuatf reviewed (commented):
Hi @hannahtinkler 👋  Thanks for the PR!

A couple quick notes before we review this PR:

* The current PR size is over 1000 lines.  It might be good to break this down into several smaller PRs so we can better review and test.  I like to shoot for 300 lines, but there isn't a hard and fast rule.
* Testing instructions need to be a bit more thorough.  I see "Make a request to the API to ensure this works as expected" - can you provide examples of requests that need to be made to any and all endpoints?

I'll be AFK for a couple days starting tomorrow.  Happy to review on Monday, but if you need a faster review, you may check with someone else on @woocommerce/flux.

hannahtinkler committed:
Reduces pull request size.

hannahtinkler commented:
Thanks @joshuatf - the PR has been updated 🙂

Konamiman commented on plugins/woocommerce/tests/php/src/Internal/PushNotifications/PushNotificationsTest.php:
```diff
@@ -0,0 +1,189 @@
+	 */
+	public function test_it_caches_enablement_state_correctly() {
+		/**
+		 * First instance with Jetpack disconnected - should return false.
```
For single-line comments we often use the `// comment` syntax instead, even if a blank line is leave between the comment and the code to preserve readability.

Konamiman commented on plugins/woocommerce/tests/php/src/Internal/PushNotifications/PushNotificationsTest.php:
```diff
@@ -0,0 +1,189 @@
+			->willReturnCallback(
+				function ( $feature_id ) use ( $feature_enabled ) {
+					/**
+					 * Return the specified value for push_notifications, false for all others.
```
I think this comment is superfluous, the code itself already reads pretty well.

Konamiman commented on plugins/woocommerce/src/Internal/PushNotifications/PushNotifications.php:
```diff
@@ -0,0 +1,90 @@
+			if (
+				class_exists( JetpackConnectionManager::class )
+				&& $proxy->get_instance_of( JetpackConnectionManager::class )->is_connected()
+			) {
```
Could we use a ternaty operator here? As in: `$this->enabled = (class_exists(...) && $proxy->...)`

Konamiman commented on plugins/woocommerce/src/Internal/PushNotifications/PushNotifications.php:
```diff
@@ -0,0 +1,90 @@
+				$this->enabled = false;
+			}
+		} catch ( Exception $e ) {
+			$this->enabled = false;
```
It makes sense to leave the class disabled in case of error. But it would be good to also log the error for later troubleshooting any issue.

Hints for unit testing this:

1. Use a new mock of `JetpackConnectionManager` that will thrown an error on the `is_connected` method.
2. Use `LegacyProxy` to run `wc_get_logger`, and in the test use a fake logger class that just logs to a local array. Search for instances of `$fake_logger` in the WooCommerce code for example usages of this strategy.

Konamiman reviewed (changes requested):
Hi @hannahtinkler, I left some nitpick comments but I like to see that you added unit tests from the very beginning and use `LegacyProxy`, mock classes and container replacements as intended.

However, while initially this pull request was too big, I feel like perhaps it's now too small, as all the added class does is checking if it should be enabled or not. Do you think it would be possible to add some of the intended functionality, even if very minimally, without making the pull request too big again? If there's no reasonable way to do that we can leave it as is, but I wanted to ask first. Happy to discuss options in further comments before you do any actual change.

hannahtinkler committed:
Add error logging to PushNotifications enablement check.

- Log exceptions when Jetpack connection check fails
- Use LegacyProxy for wc_get_logger() to enable test mocking
- Add unit test for error logging with fake logger
- Convert single-line in-method comments from /** */ to // style
- Simplify enablement logic with ternary expression

hannahtinkler committed:
Disables lint check for PHPdocs on fake logger.

hannahtinkler committed:
Adds @SInCE annotations.
